src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,130 @@
+# Uninitialized memory
+
+> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
+>
+> — _Emily Dickinson_
+
+While we have covered [invalid values], there's another thing that is a kind of invalid value, but has nothing to do with actual bit patterns: Uninitialized memory.
+
+An easy way to think about uninitialized memory is that there's an additional value that does not map to any concrete bit pattern, but can be introduced in the abstract machine in various ways, and makes _most_ values invalid.
+
+If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be constructed with no overhead and then written to in parts.
```

Comment:
I would probably explain uninitialized memory along these lines:

 1. Uninitialized memory behaves exactly like a 257th value for bytes.
 2. You cannot treat an uninitialized byte as "this is just a byte with one of the 256 normal values picked randomly". Any attempt to figure out "what the byte actually is" is UB.
 3. Some types have invalid values, and uninitialized bytes are not allowed for most types.
 4. The `MaybeUninit<T>` type allows any sequence of bytes, including uninitialized ones. The padding of structs also allows any sequence of bytes, including uninitialized ones.
 5. If you copy bytes from one place to another with `std::ptr::copy` (which is _untyped_), then this perfectly copies the bytes, including whether they are uninitialized.
 6. If you copy bytes using a _typed copy_, then this promises fewer things than an untyped copy: it will only preserve things relevant for that type. For one, it is UB if the source bytes are invalid for the type. For another, if the type is a struct, then padding is not copied — instead, uninitialized bytes are written to the destination even if the source has initialized the padding (of course, the compiler might _implement_ this by just copying the padding, but it also might not). 
 7. Memory is uninitialized when it comes from the allocator. When the memory is freed, this is worse than just being uninitialized — it is UB to read it even as an `MaybeUninit` or with an untyped copy.

Reply:
I've expanded this a bunch, thanks!